### PR TITLE
Added variants of A*

### PR DESCRIPTION
**Pull request by sarimmehdi**
Implemented A* with beam search, iterative deepening, dynamic weighting, theta star, and jump point

### CODE REVIEW

**Comment by sarimmehdi** [comment]
Demo: https://www.youtube.com/watch?v=86ST5PbcuBs

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for interesting PR. I have some minor comments before deep review. PTAL.

**Comment by sarimmehdi** [comment]
@AtsushiSakai sorry for the late reply. I have made the requested changes. Please have a look.

**Comment by sarimmehdi** [comment]
@AtsushiSakai `Build execution time has reached the maximum allowed time for your plan (60 minutes)` I don't know how to deal with this. But everything else seems to be working fine. Please let me know if it is okay with you.

**Comment by sarimmehdi** [comment]
@AtsushiSakai I think everything is fine now, maybe you can have a look at it.

File: test_a_star_variants.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import PathPlanning.AStar.a_star_variants as astar
```
Comment (AtsushiSakai) [inline test_a_star_variants.py:1]: Could you please remove this file? it seems duplicate.

File: PathPlanning/AStar/a_star_variants.py [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+            counter += 1
+            if [x + i_temp, y + j_temp] == self.goal_pt:
+                got_goal = True
+                break
+            if (x + i_temp, y + j_temp) not in self.obs_grid.keys():
+                break
+        return i_temp - 2*i, j_temp - 2*j, counter, got_goal
+
+    def jump_point(self):
+        '''Jump point: Instead of exploring all empty spaces of the
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/a_star_variants.py:177]: Could you please change ''' to """ ?

File: PathPlanning/AStar/a_star_variants.py [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+            current_node['open'] = False
+            current_node['in_open_list'] = False
+            plt.plot(current_node['pos'][0], current_node['pos'][1], "g*")
+            del self.open_set[p]
+            current_node['fcost'], current_node['hcost'] = np.inf, np.inf
+        if show_animation:
+            plt.show()
+
+    def astar(self):
+        '''Beam search: Maintain an open list of just 30 nodes.
```
Comment (AtsushiSakai) [inline PathPlanning/AStar/a_star_variants.py:260]: Could you please change ''' to """ ?

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]

**Review by AtsushiSakai — CHANGES_REQUESTED** [review]
Thank you for fixing. This is the last review. After fixing these, I will merge it. PTAL.

**Comment by AtsushiSakai** [comment]
I have a question. When I use Jump Point search option, It looks like final path is collided to some obstacles. Is it expected?
SS 2020-09-20 at 13 15 12

**Comment by sarimmehdi** [comment]
@AtsushiSakai I implemented the requested changes. For jump-point I forgot to change the parameter for the smaller map. I apologize for the oversight. This is what the path should look like:

Zoomed-out (it might seem like the path is crossing the obstacle but once you zoom in, you see that is not the case):
jump_point_zoomed_out

Zoomed-in (no path crosses the obstacle):
jump_point

**Review by AtsushiSakai — APPROVED** [review]
LGTM. Thank you so much for great PR!.

**Pull request merged**